aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -656,6 +656,59 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
   return at::index(self, indices);
 }
 
+Tensor _unsafe_masked_index(const Tensor& self, const Tensor& mask, const torch::List<c10::optional<Tensor>>& indices, const Scalar& fill) {
+  // Unsafe masked index is equivalent to
+  //   where(mask, self[indices], fill)
+  // with the main difference being that the when the `mask` is false, the tensor
+  // `self` is not indexed using `indices`. This allows `indices` to be out-of-bounds
+  // when `mask` is false. When `mask` is true, the `indices` are expected to be
+  // in bounds and is not checked.
+  //
+  // This function is not meant to be executed on eager mode. An unoptimized version
+  // is provided here.
+  //
+  // compiler backends should implement this op such that `self[indices]` is not
+  // loaded when `mask` is true. See inductor for a reference.
+  auto clamp = [](const c10::optional<Tensor>& index, auto size) -> c10::optional<Tensor> {
+    if (!index) {
+      return index;
+    }
+    // Disallow bool
+    auto dtype = index->scalar_type();
+    TORCH_CHECK(dtype == kLong || dtype == kInt,
+                "_unsafe_masked_index found unexpected index type ", dtype);
+    return at::clamp(*index, -size, size - 1);
```

Comment:
Hazard here for empty inputs:
```
2024-04-11T21:17:58.2785208Z IndexError: index is out of bounds for dimension with size 0
```

It may be reasonable to just return a `full` tensor when there are no valid indices. It may be non-trivial to determine the correct output shape though.